src/solvers.jl

```diff
@@ -31,31 +39,69 @@ function solve(m::Model; suppress_warnings=false, ignore_solve_hook=(m.solvehook
         return solvenlp(m, suppress_warnings=suppress_warnings)
     end
 
-    if isa(m.solver,UnsetSolver) &&
-      (length(m.obj.qvars1) > 0 || length(m.quadconstr) > 0)
-        m.solver = MathProgBase.defaultQPsolver
-    end
-    if anyInts
-        if isa(m.solver,UnsetSolver)
-            m.solver = MathProgBase.defaultMIPsolver
-            s = solveMIP(m; suppress_warnings=suppress_warnings)
-            # Clear solver in case we change problem types
-            m.solver = UnsetSolver()
-            m.internalModelLoaded = false
-            return s
-        else
-            solveMIP(m; suppress_warnings=suppress_warnings)
+    unset = m.solver == UnsetSolver()
+    traits = problemclass(m)
+    buildInternalModel(m, traits, suppress_warnings=suppress_warnings)
+
+    MathProgBase.optimize!(m.internalModel)
+    stat = MathProgBase.status(m.internalModel)
+
+    if stat == :Optimal
+        if !(traits.int | traits.sos | traits.sdp)
+            m.redCosts = try
+                MathProgBase.getreducedcosts(m.internalModel)
+            catch
+                fill(NaN, length(m.colVal))
+            end
+
+            m.linconstrDuals = try
+                MathProgBase.getconstrduals(m.internalModel)
+            catch
+                fill(NaN, length(m.linconstr))
+            end
         end
     else
-        if isa(m.solver,UnsetSolver)
-            m.solver = MathProgBase.defaultLPsolver
-            s = solveLP(m, suppress_warnings=suppress_warnings)
-            m.solver = UnsetSolver()
-            return s
-        else
-            solveLP(m; suppress_warnings=suppress_warnings)
+        suppress_warnings || warn("Not solved to optimality, status: $stat")
+
+        if stat == :Infeasible
+            m.linconstrDuals = try
+                MathProgBase.getinfeasibilityray(m.internalModel)
+            catch
+                suppress_warnings || warn("Infeasibility ray (Farkas proof) not available")
+                fill(NaN, length(m.linconstr))
+            end
+        elseif stat == :Unbounded
+            try
+                m.colVal = MathProgBase.getunboundedray(m.internalModel)
+            catch
+                suppress_warnings || warn("Unbounded ray not available")
+            end
+        end
+    end
+
+    m.objVal = NaN
+    m.colVal = fill(NaN, m.numCols)
+    if !(stat == :Infeasible || stat == :Unbounded)
+        try
+            objVal = MathProgBase.getobjval(m.internalModel) + m.obj.aff.constant
+            colVal = MathProgBase.getsolution(m.internalModel)
+            m.objVal = objVal # Don't corrupt the answers if one of the above two calls fails
+            m.colVal = colVal
+        end
+    end
+
+    if traits.sdp && m.objSense == :Max
```

Comment:
Not ideal that the logic for the objective flip leaks through here, but not a big deal